Dev gear 1425

neurosutton opened the pull request:
Please review the overall strategy of:
(1) writing config options based on the BIDS curation after inside the gear and 
(2) passing the config along to the other "gears" that have been subsumed into this monster gear.

This stage is not considered functional in the least, but is a massive step that needs to be approved or reimagined.

Main files to consider for this review are:
run.py
utils.bids.bids_file_locator

neurosutton committed:
Adapt manifest for BIDS = no inputs

neurosutton committed:
Init commit to transform to BIDS

neurosutton committed:
Add BIDS error checking options

neurosutton committed:
Metamorphize structural gear into separate folder

neurosutton committed:
Add beginnings of poetry

neurosutton committed:
Add func and diff input options

neurosutton committed:
Erect a skeleton

neurosutton committed:
Add more notes

AndysWorth commented on fw_gear_hcp_struct/args/PreFreeSurfer.py:
```diff
@@ -125,7 +128,10 @@ def build(context):
+            if not params["echodiff"]:
+                # In the case of merged magnitude files, echodiff is already calculated correctly
+                # in bids_file_locator.define_bids_files
+                params["echodiff"] = format((echotime2 - echotime1) * 1000.0, ".15f")
```
Nice!

AndysWorth commented on manifest.json:
```diff
@@ -1,19 +1,19 @@
 {
-  "name": "hcp-struct",
+  "name": "bids_hcp_struct",
```
Should this be "bids_hcp"?

neurosutton replied:
Yes, thank you.

AndysWorth commented on pyproject.toml:
```diff
@@ -0,0 +1,19 @@
+[tool.poetry]
+name = "bids_hcp_struct"
```
should this be "bids_hcp"?

AndysWorth commented on pyproject.toml:
```diff
@@ -0,0 +1,19 @@
+[tool.poetry]
+name = "bids_hcp_struct"
+version = "0.0.1"
+description = "Structural processing according to HCP pipeline methodology"
```
Just "hcp" not structural?

neurosutton replied:
Yep, thank you.

AndysWorth commented on utils/bids/bids_file_locator.py:
```diff
@@ -0,0 +1,248 @@
+            # use and check for errors.
+            # Error logging handled within validate_bids.
+
+            # Need to convert from cli to python bids_validator (npm unstable solution)
```
Last I looked, the python BIDS validator only validates individual files while the regular one goes through everything.  I'm not sure that makes a difference.  In any event, this can be dealt with later.

neurosutton replied:
On the todo list. Thank you.

AndysWorth commented on utils/bids/bids_file_locator.py:
```diff
@@ -0,0 +1,248 @@
+        config = self.gtk_context.config
+        try:
+            log.info('Downloading BIDS')
+            self.gtk_context.download_session_bids(target_dir=self.gtk_context.work_dir / "bids")
```
Don't you want something like [download_bids_for_runlevel()](https://gitlab.com/flywheel-io/flywheel-apps/templates/bids-app-template/-/blob/main/utils/bids/download_run_level.py) here?  I would think it needs to load all data for a subject (all sessions).

neurosutton replied:
Yes - I still have things to learn about the SDK and GearToolkit. Thanks!

AndysWorth reviewed (approved):
Should this be on GitLab?

I'm assuming that most of the un-touched files are just copied from their respective parts so I didn't go through them.

I went through this fairly quickly but I think it should work, assuming it's not too big to assemble.

neurosutton commented:
Right - I will migrate this to GitLab. There are likely a fair number of things in the untouched files that will need to be checked for flexibility, but by and large, I'm hoping to leave them alone. Thanks for your input on such short notice, Andy!

neurosutton committed:
Fix struct refs and format

naterichman commented on README.md:
```diff
@@ -36,6 +39,65 @@
+2. Field map for correcting readout distortion
+    * Option 1: GRE = "typical" GRE B0 field map including magnitude and phase volumes
+    * Option 2: SpinEchoFieldMap = a pair of spin echo with opposite phase-encode directions ("Positive" = R>>L or P>>A, and "Negative" = L>>R or A>>P) for "TOPUP"-style distortion estimation
+3. StructZip output from the HCP-Struct gear (containing <code>T1w/</code>, <code>T2w/</code>, and <code>MNINonLinear/</code> folders)
```
Just a note, to use \`\` instead of `<code></code>` tags

naterichman commented on fw_gear_hcp_diff/args/DiffPreprocPipeline.py:
```diff
@@ -0,0 +1,383 @@
+"""
+Builds, validates, and excecutes parameters for the HCP script
+/opt/HCP-Pipelines/DiffusionPreprocessing/DiffPreprocPipeline.sh
```
Just to confirm this is not written by us, this is supplied by the HCP?

neurosutton replied:
I believe so.

naterichman commented on fw_gear_hcp_diff/args/DiffPreprocPipeline.py:
```diff
@@ -0,0 +1,383 @@
+log = logging.getLogger(__name__)
+
+
+def build(context):
```
I would encourage typing here, just for readability.  I think this is a gear context, but it would be helpful to have either 
* `def build(context: flywheel_gear_toolkit.GearToolkitContext) -> None`
* A docstring such as:
```python
def build(context):
"""Builds context for script.

Args:
    context (GearToolkitContext)

Returns
    (None)
"""
```

naterichman commented on fw_gear_hcp_diff/args/DiffPreprocPipeline.py:
```diff
@@ -0,0 +1,383 @@
+
+def build(context):
+    config = context.config
+    inputs = context._invocation["inputs"]
```
Any reason to not use `context.inputs`?

neurosutton replied:
Old code... trying to leave virtually everything alone that I can, so there is less to have break from currently functional.

naterichman commented on fw_gear_hcp_diff/args/DiffPreprocPipeline.py:
```diff
@@ -0,0 +1,383 @@
+    context.gear_dict["Diff-params"] = params
+
+
+def validate(context):
```
High level here, I would consider reorganizing here, chunking smaller portions together and combining validate and build.  Right now you're getting the same data twice, once for validating, and once for building.  Instead consider chunking into sections of building+validating, e.g. 

```python
def process_DWI_data(inputs: t.Dict[str,str]):
    """Handle DWIPositiveData and DWINegativeData validation
            and populate section of the /opt/HCP-Pipelines/DiffusionPreprocessing/DiffPreprocPipeline.sh
            script
    """
    # code

# Later
def validate_and_build_diff_preproc(context: GearToolkitContext):
    process_DWI_data(context.inputs)
    # more handlers
```

neurosutton replied:
As I've looked through the comments and tried to address some, I'm beginning to abandon the idea of keeping this gear very similar to the original coding. I will look at this optimization specifically as I keep the main ideas, steps, and commands sent to HCP, but reimagine the functions that get there. Thank you.

naterichman commented on fw_gear_hcp_diff/args/DiffPreprocPipeline.py:
```diff
@@ -0,0 +1,383 @@
+            pedir_pos = tr("ijk", "xyz", pedir_pos)
+            pedir_neg = tr("ijk", "xyz", pedir_neg)
+            if pedir_pos == pedir_neg:
+                raise Exception(
```
NIT: suggest raising `ValueError` instead

naterichman commented on fw_gear_hcp_diff/diff_main.py:
```diff
@@ -0,0 +1,152 @@
+
+    try:
+        # Build and validate from Volume Processing Pipeline
+        DiffPreprocPipeline.build(context)
```
I would avoid setting values in the context itself and instead have this function return a list of params.

naterichman commented on fw_gear_hcp_diff/scripts/hcpdiff_qc_mosaic.sh:
```diff
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+set -e
```
Yeaaaa I'm not a big bash script fan so I'm going to skip this one for now :)

naterichman commented on manifest.json:
```diff
@@ -139,7 +71,79 @@
+    },
+    "gear-abort-on-bids-error": {
+        "type": "boolean",
+        "default": false,
```
Why default to false?

neurosutton replied:
Mainly b/c I wasn't sure where all this would get carried through and I wanted the script to be able to produce errors. Definitely makes more sense to be true after more dev!

naterichman commented on utils/bids/bids_file_locator.py:
```diff
@@ -0,0 +1,334 @@
+
+
+class bidsInput:
+    def __init__(self, pth=None):
+        if pth:
+            self.gtk_context = GearToolkitContext(config_path=pth)
+        else:
+            self.gtk_context = GearToolkitContext()
```
Why not just:

```suggestion
    def __init__(self, context: GearToolkitContext):
        self.gtk_context = context
```

naterichman commented on utils/bids/run_level.py:
```diff
@@ -0,0 +1,109 @@
+        This function is meant to be run by analysis gears.
+    """
+
+    try:
```
There are a few improvements that could happen here, I'd be happy to help optimize this with you, but since it's only run once per gear run, it's probably not super important to optimize.

neurosutton committed:
Update more fields

naterichman commented on fw_gear_hcp_diff/diff_main.py:
```diff
@@ -0,0 +1,152 @@
+    # Preamble: take care of all gear-typical activities.
+    context.gear_dict = {}
+    # Initialize all hcp-gear variables.
+    gear_preliminaries.initialize_gear(context)
```
Can't find this function,.

neurosutton replied:
It comes from hcp-base, a private gear, that creates the base Docker image and populates utils with more functions.

neurosutton replied:
I am going to pull the "shared" modules directly into this utils directory. Some of them look like they are completely superseded by GearToolkitContext calls.

naterichman reviewed (changes requested):
I'm not sold on the overally strategy of tricking the gears into having config's like if the user ran them themselves.  I totally understand the reason for going there since this is on a tight-ish deadline, but I don't think it's best practice.

Instead, I would much rather see each gears' `run` function take in the specific input and configs that are necessary for that gear.  For example, right now `diff_main.run` takes a `context` as an argument, I would love to see it take a dictionary of inputs, and their location, as well as the specific configuration it needs, freesurfer license, etc.

At this point, I have very little insight into the requirements of when this gear is delivered and some of it will come down to that.  If this approach works and allows for faster release of the gear (and thats needed), then I think we have to do that.  But I strongly suggest the alternative method if we have the time.

Another possibility (even more work) @AndysWorth is to release initial version with this approach working, as we start to refactor the HCP gears, we make them compatible with the new template Nicolas and I have been working on, and allow them to be pip installable.  Right now it would be hard to maintain any changes to any of these components since it would need to be changed in the standalone gear AND this one.  The pip-installed approach would allow this gear to be an "orchestrator" handling the bids portion and pass the required information into the sub-components.

neurosutton commented:
I like the idea of an orchestrator gear, though probably not feasible with the time crunch. Just to alleviate one concern about maintaining - the instructions for the standalone gears are basically, "No touchie!" :)

neurosutton committed:
Remove debug option

Co-authored-by: Nathan Richman <68390038+naterichman@users.noreply.github.com>

neurosutton committed:
Update README

neurosutton committed:
Remove boilerplate and custom log
Import GearToolkitContext.interfaces.command_line to handle
the build, validate, and execute commandline items
Change to logging, per standard practices

neurosutton committed:
Implement dictionary style gear_args

naterichman reviewed (approved):
I like this approach better! Not directly manipulating the GearToolkitContext works better I think!

neurosutton committed:
Docker and organizational overhaul

neurosutton commented:
Merging and then migrating to GitLab
